src/tests/test_upgrade_agent.py

```diff
@@ -0,0 +1,189 @@
+    return stdout.decode("utf-8")
+
+
+def image_repo(image: str) -> str:
+    """
+    Calculates the image repository from the given image reference. For example, if the image is
+    'quay.io/my/image:v1.23' it will return 'quay.io/my/image'.
+    """
+    return re.sub(r":.*$", "", image)
+
+
+def image_digest(image: str) -> str:
+    """
+    Gets the digest of the given image.
+    """
+    result = skopeo(["inspect", "--raw", f"docker://{image}"])
+    result = json.loads(result)
+    return result["config"]["digest"]
+
+
+def repo_tags(repo: str) -> str:
+    """
+    Gets the lit of tags of the given image respository"
+    """
+    result = skopeo(["list-tags", f"docker://{repo}"])
+    result = json.loads(result)
+    return result["Tags"]
+
+
+def parse_tag(tag: str) -> List[int]:
+    """
+    Converts a tag like 'v1.2.3-45' into a list of 4 integers like [1, 2, 3, 45]. This is intended for later
+    sorting lexicographically. Returns None if the given tag isn't a version number.
+    """
+    match = re.match(r"^v(?P<x>\d+).(?P<y>\d+).(?P<z>\d+)(-(?P<r>\d+))?$", tag)
+    if match is None:
+        return None
+    x = int(match.group("x"))
+    y = int(match.group("y"))
+    z = int(match.group("z"))
+    r = match.group("r")
+    if r is not None:
+        r = int(r)
+    else:
+        r = 0
+    return [x, y, z, r]
+
+
+class TestUpgradeAgent(BaseTest):
+    @JunitTestSuite()
+    def test_upgrade_agent(self, cluster):
+        """
+        This test will prepare the cluster with the previouis agent image. Once it is ready to install it will restart
+        the service with an agent image that doesn't exist, so that all the hosts will move to `insufficient`. Then it
+        will restart the service again with the current image, so that host will upgrade and move to `ready` again.
+        """
+        # Find the namespace:
+        namespace = os.environ.get("NAMESPACE", "assisted-installer")
+
+        # Create the API client:
+        client = kubernetes.dynamic.DynamicClient(
+            kubernetes.client.api_client.ApiClient(
+                configuration=kubernetes.config.load_kube_config(),
+            ),
+        )
+        configmaps_api = client.resources.get(api_version="v1", kind="ConfigMap")
+        deployments_api = client.resources.get(api_version="apps/v1", kind="Deployment")
+
+        # Get the digest and tags of the agent image currently used by the service:
+        service_config = configmaps_api.get(namespace=namespace, name="assisted-service-config")
+        current_image = service_config.data["AGENT_DOCKER_IMAGE"]
+        agent_repo = image_repo(current_image)
+
+        # We expect the repository to contain tags like 'v2.3.1'. We want to find the newest of those tags that is
+        # actually different (different digest) than the current tag used by the service, to make sure that we are using
+        # two different images.
+        all_tags = repo_tags(agent_repo)
```

Comment:
some part of the following logic can be refactored into a method, to abstract-away some of the more technical things here

Reply:
Done.